src/main/java/com/microsoft/graph/serializer/FallbackTypeAdapterFactory.java

```diff
@@ -89,16 +92,70 @@ public FallbackTypeAdapterFactory(@Nonnull final ILogger logger) {
     public <T> TypeAdapter<T> create(@Nonnull final Gson gson, @Nonnull final TypeToken<T> type) {
         Objects.requireNonNull(type, "parameter type cannot be null");
         final Class<T> rawType = (Class<T>) type.getRawType();
+
         if (rawType.isEnum()) {
             return new EnumTypeAdapter<T>(rawType, logger);
         } else if (rawType == Void.class) {
             return (TypeAdapter<T>) voidAdapter;
+        } else if (IJsonBackedObject.class.isAssignableFrom(type.getRawType())) {
+
+            // Avoid overriding custom IJsonBackedObject type adapters defined in GsonFactory
+            if (BaseCollectionResponse.class.isAssignableFrom(rawType) || BaseCollectionPage.class.isAssignableFrom(rawType)) {
+                return null;
+            }
+
+            final TypeAdapter<?> delegatedAdapter = gson.getDelegateAdapter(this, type);
+            return (TypeAdapter<T>) new ODataTypeParametrizedIJsonBackedObjectAdapter(gson, delegatedAdapter, type);
         }
         else {
             return null;
         }
     }
 
+    /**
+     * This adapter is responsible for deserialization of IJsonBackedObjects where service
+     * returns one of several derived types of a base object, which is defined using the
+     * odata.type parameter. If odata.type parameter is not found, the Gson default
+     * (delegated) type adapter is used.
+     */
+    private class ODataTypeParametrizedIJsonBackedObjectAdapter extends TypeAdapter<IJsonBackedObject> {
+
+        private final Gson gson;
+        private final TypeAdapter<?> delegatedAdapter;
+        private final TypeToken<?> type;
+
+        public ODataTypeParametrizedIJsonBackedObjectAdapter(@Nonnull Gson gson, @Nonnull TypeAdapter<?> delegatedAdapter, @Nonnull final TypeToken<?> type) {
+            super();
+            this.gson = gson;
+            this.delegatedAdapter = delegatedAdapter;
+            this.type = type;
+        }
+
+        @Override
+        public void write(JsonWriter out, IJsonBackedObject value)
+            throws IOException
+        {
+            ((TypeAdapter<IJsonBackedObject>)this.delegatedAdapter).write(out, value);
+        }
+
+        @Override
+        public IJsonBackedObject read(JsonReader in) {
+            JsonElement jsonElement = Streams.parse(in);
+
+            if (jsonElement.isJsonObject()) {
+                final DerivedClassIdentifier derivedClassIdentifier = new DerivedClassIdentifier(logger);
```

Comment:
maybe have this as a property assigned by constructor to avoid assigning at each method call?